tests/torchtune/utils/test_distributed.py

```diff
@@ -411,28 +421,86 @@ def test_qlora_state_dict(self):
             "max_seq_len": 64,
             "lora_rank": 4,
             "lora_alpha": 1.0,
+            "quantize_base": True,
         }
+        # single-device model as groundtruth
         with torch.device("cuda"):
-            lora_kwargs = dict({"quantize_base": False}, **kwargs)
-            model_lora = lora_llama2(**lora_kwargs)
-        full_sd = model_lora.cpu().state_dict()
+            base_model = lora_llama2(**kwargs)
+        set_trainable_params(base_model, get_adapter_params(base_model))
+        if enable_activation_checkpointing:
+            utils.set_activation_checkpointing(
+                base_model, auto_wrap_policy={modules.TransformerDecoderLayer}
+            )
+
+        # fsdp model for saving state dict
+        fsdp_model_to_save = copy.deepcopy(base_model)
+        for m in fsdp_model_to_save.modules():
+            if enable_activation_checkpointing:
+                if isinstance(m, CheckpointWrapper):
+                    fully_shard(m)
+            else:
+                if isinstance(m, modules.TransformerDecoderLayer):
+                    fully_shard(m)
+        fully_shard(fsdp_model_to_save)
```

Comment:
The purpose here is to recreate the sharding of the QLoRA model, right? If so, why not shard LoRA layers first as is done elsewhere?

Reply:
for exact QLoRA setup, I was actually thinking of using recipe test as the best place

for unit test and for best QPS, wrapping `TransformerDecoderLayer` is actually more optimal. I was wrapping LoRA to be fair with FSDP1 recipe. FSDP1 is base on flat-parameter and it has to wrap LoRA otherwise it cannot handle a mixing freezed and trainable parameters in one flat-parameter